Add blog about docker hub rate limit (#8627)
istio/istio.io: content/en/blog/2020/docker-rate-limit/index.md

---
title: Handling Docker Hub rate limiting
description: How to ensure your clusters are not impacted by Docker Hub rate limiting.

> publishdate: 2020-12-05

Reviewer: ```suggestion
publishdate: 2020-12-07
```

attribution: John Howard (Google)
keywords: [docker]
target_release: 1.8
---

Since November 20th, 2020, Docker Hub has introduced [rate limits](https://www.docker.com/increase-rate-limits) on image pulls.

Because Istio uses [Docker Hub](https://hub.docker.com/u/istio) as the default registry, usage on a large cluster may lead

> to pods failing to startup due to exceeding rate limits. This can be especially problematic for Istio, as there is typically

Reviewer 2: If you have an error message it can be useful for users to identify the underlying problem.

Author: good idea, I can add it, its in the dockerhub link above as well

the Istio sidecar image alongside most pods in the cluster.

> # Mitigations

Reviewer: A couple of lint errors still. You need to add ECR to .spelling and:
```suggestion
## Mitigations
```

> Istio allows using a custom docker registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.

Reviewer 2: ```suggestion
Istio allows providing a custom docker registry which can be used so that the images can be fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.
```

> Istio allows providing a custom docker registry which can be used so that the images can be fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.

Reviewer 2: ```suggestion
Istio allows providing a custom docker registry which can be used so that the images are fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.
```

> Istio allows providing a custom docker registry which can be used so that the images can be fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.

Reviewer: ```suggestion
Istio allows you to specify a custom docker registry which you can use to make container images be fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.
```

> Istio allows you to specify a custom docker registry which you can use to make container images be fetched from your private registry. This can be configured by passing `--set hub=<some-custom-registry>` at installation time.

Reviewer 3: This seems to imply a particular installation methodology. Is this instruction correct for an operator-based install, for example?

Author: We don't really have a consistent way to explain how to set an installation option. I have https://github.com/istio/istio.io/issues/4945 opened, but its quite old, to have a common abstraction in the docs for this. From the --set command a user using the operator could derive the IstioOperator yaml, although I am not sure all users know how to do this; --set is basically the same yaml API but flattened

> Istio provides official mirrors to [Google Container Registry](`gcr.io/istio-release`). This can be configured with `--set hub=gcr.io/istio-release`. This is available for Istio 1.5+.

Reviewer: ```suggestion
Istio provides official mirrors to [Google Container Registry](gcr.io/istio-release). This can be configured with `--set hub=gcr.io/istio-release`. This is available for Istio 1.5+.
```

> Alternatively, you can copy the official Istio images to your own registry. This is especially useful if your cluster runs in an environment with a registry tailored for your environment (for example, on AWS you may want to mirror images to Amazon ECR). This can be done with the following script:

Reviewer 2: ```suggestion
Alternatively, you can copy the official Istio images to your own registry. This is especially useful if your cluster runs in an environment with a registry tailored for your use case (for example, on AWS you may want to mirror images to Amazon ECR) or you have air-gapped security requirements where access to public registries is restricted. This can be done with the following script:
```

{{< text bash >}}
$ SOURCE_HUB=istio
$ DEST_HUB=my-registry # Replace this with the destination hub
$ IMAGES=( install-cni operator pilot proxyv2 ) # Images to mirror.
$ VERSIONS=( 1.7.5 1.8.0 ) # Versions to copy
$ VARIANTS=( "" "-distroless" ) # Variants to copy
$ for image in $IMAGES; do
$ for version in $VERSIONS; do
$ for variant in $VARIANTS; do
$   name=$image:$version$variant
$   docker pull $SOURCE_HUB/$name
$   docker tag $SOURCE_HUB/$name $DEST_HUB/$name
$   docker push $DEST_HUB/$name
$   docker rmi $SOURCE_HUB/$name
$   docker rmi $DEST_HUB/$name
$ done
$ done
$ done

> {{< /text >}}

Reviewer 2: Is it better done in a Github snipper/repo/wiki so we can augment it for newer versions?

Author: I don't expect this to change and it needs manual user interaction (to set the version, dest_hub, etc), so I don't think it *needs* to - but happy to move if we think that is preferred.

Reviewer 2: It's fine to keep it here I guess.
